src/System.Buffers.Primitives/System/Collections/Position.cs

```diff
@@ -0,0 +1,82 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System.Buffers;
+using System.ComponentModel;
+using System.Diagnostics;
+using System.Runtime.CompilerServices;
+
+namespace System.Collections.Sequences
+{
+    public readonly struct Position : IEquatable<Position>
+    {
+        readonly object _segment;
+        readonly int _index;
+
+        public Position(object segment, int index)
+        {
+            _segment = segment;
+            _index = index;
+        }
+
+        public Position(object segment)
+        {
+            _segment = segment;
+            _index = 0;
+        }
+
+        public object Segment => _segment;
+        public int Index => _index;
+
+        public static explicit operator int(Position position) => position._index;
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public (T segment, int index) Get<T>()
+        {
+            var segment = _segment == null ? default : (T)_segment;
+            return (segment, _index);
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public T GetSegment<T>()
+        {
+            switch (Segment)
+            {
+                case null:
+                    return default;
+                case T segment:
+                    return segment;
+            }
+
+            ThrowHelper.ThrowInvalidOperationException(ExceptionResource.UnexpectedSegmentType);
+            return default;
+        }
+
+        public static bool operator ==(Position left, Position right) => left._index == right._index && left._segment == right._segment;
+        public static bool operator !=(Position left, Position right) => left._index != right._index || left._segment != right._segment;
+
+        public static Position operator +(Position value, int index)
```

Comment:
I thought we were going to limit the ability to modify positions?

Reply:
We now have public ctors, Index, and Segment. The + operator is much more hidden than those. 
Also, without the operator, code that uses position as index is very wacky. It does not care about the Segment, yet it has to deal with it (in the ctor, at least).

We can think about it more, but I would treat this as a separate PR.